charts/nri-kube-events/templates/_helpers_compatibility.tpl

```diff
@@ -3,8 +3,354 @@ Returns a dictionary with legacy runAsUser config.
+{{- $oldRegistry := include "nri-kube-events.compatibility.old.integration.registry" . -}}
+{{- $newRegistry := include "nri-kube-events.compatibility.new.integration.registry" . -}}
+{{- $registry := $oldRegistry | default $newRegistry | default $globalRegistry -}}
+
+{{- $oldRepository := include "nri-kube-events.compatibility.old.integration.repository" . -}}
+{{- $newRepository := include "nri-kube-events.compatibility.new.integration.repository" . -}}
+{{- $repository := $oldRepository | default $newRepository }}
+
+{{- $oldTag := include "nri-kube-events.compatibility.old.integration.tag" . -}}
+{{- $newTag := include "nri-kube-events.compatibility.new.integration.tag" . -}}
+{{- $tag := $oldTag | default $newTag -}}
+
+{{- if $registry -}}
+    {{- printf "%s/%s:%s" $registry $repository $tag -}}
+{{- else -}}
+    {{- printf "%s:%s" $repository $tag -}}
+{{- end -}}
+{{- end -}}
+
+
+
+{{/*
+Creates the image string needed to pull the agent's image respecting the breaking change we made in the values file
+*/}}
+{{- define "nri-kube-events.compatibility.images.agent" -}}
+{{- $globalRegistry := include "nri-kube-events.compatibility.global.registry" . -}}
+{{- $oldRegistry := include "nri-kube-events.compatibility.old.agent.registry" . -}}
+{{- $newRegistry := include "nri-kube-events.compatibility.new.agent.registry" . -}}
+{{- $registry := $oldRegistry | default $newRegistry | default $globalRegistry -}}
+
+{{- $oldRepository := include "nri-kube-events.compatibility.old.agent.repository" . -}}
+{{- $newRepository := include "nri-kube-events.compatibility.new.agent.repository" . -}}
+{{- $repository := $oldRepository | default $newRepository }}
+
+{{- $oldTag := include "nri-kube-events.compatibility.old.agent.tag" . -}}
+{{- $newTag := include "nri-kube-events.compatibility.new.agent.tag" . -}}
+{{- $tag := $oldTag | default $newTag -}}
+
+{{- if $registry -}}
+    {{- printf "%s/%s:%s" $registry $repository $tag -}}
+{{- else -}}
+    {{- printf "%s:%s" $repository $tag -}}
+{{- end -}}
+{{- end -}}
+
+
+
+{{/*
+Returns the pull policy for the integration image taking into account that we made a breaking change on the values path.
+*/}}
+{{- define "nri-kube-events.compatibility.images.pullPolicy.integration" -}}
+{{- $old := include "nri-kube-events.compatibility.old.integration.pullPolicy" . -}}
+{{- $new := include "nri-kube-events.compatibility.new.integration.pullPolicy" . -}}
+
+{{- $old | default $new -}}
+{{- end -}}
+
+
+
+{{/*
+Returns the pull policy for the agent image taking into account that we made a breaking change on the values path.
+*/}}
+{{- define "nri-kube-events.compatibility.images.pullPolicy.agent" -}}
+{{- $old := include "nri-kube-events.compatibility.old.agent.pullPolicy" . -}}
+{{- $new := include "nri-kube-events.compatibility.new.agent.pullPolicy" . -}}
+
+{{- $old | default $new -}}
+{{- end -}}
+
+
+
+{{/*
+Returns a merged list of pull secrets ready to be used
```

Comment:
Since the common library accept [a list](https://github.com/newrelic/helm-charts/blob/7166933861a53a3f766083df96d8f7fe67aa5cc9/library/common-library/templates/_images.tpl#L58), couldn't we pass both the old and the new imagePullSecret?

Is there a any drowback?

Reply:
For some reason, if a template returns something like
```yaml
- test: test
- test2: test2
```
Is not able un unmarshal it to
```json
[
  {"test": "test"},
  {"test2":"test2"}
]
```

I am getting these errors:
```
error unmarshaling JSON: while decoding JSON: json: cannot unmarshal array into Go value of type map[string]interface {}
```

Reply:
```mustache
{{- define "test.helper.printsListOfObjects" -}}
- test: test
- test2: test2
{{- end -}}

{{- define "test.helper.fromToYaml" -}}
{{- include "test.helper.printsListOfObjects" . | fromYaml | toYaml -}}
{{- end -}}
```
```yaml
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: test
data:
  test.helper.printsListOfObjects: |
    {{- include "test.helper.printsListOfObjects" . | nindent 4 }}
  test.helper.fromToYaml: |
    {{- include "test.helper.fromToYaml" . | nindent 4 }}
```
`helm template test test -s templates/test-configmap.yaml` prints:
```
---
# Source: test/templates/test-configmap.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: test
data:
  test.helper.printsListOfObjects: |
    - test: test
    - test2: test2
  test.helper.fromToYaml: |
    Error: 'error unmarshaling JSON: while decoding JSON: json: cannot unmarshal array
      into Go value of type map[string]interface {}'
```

Reply:
As the error suggest, not arrays are not allowed, it seems not all valid yaml can be parsed with [fromYaml](https://github.com/helm/helm/blob/main/pkg/engine/funcs.go#L92).

Reply:
```
{{/*
Returns a merged list of pull secrets ready to be used
*/}}
{{- define "nri-kube-events.compatibility.images.pullSecrets" -}}

{{- $list := list -}}

{{- if .Values.image -}}
    {{- if .Values.image.pullSecrets -}}
        {{- $list = append $list .Values.image.pullSecrets }}
    {{- end -}}
{{- end -}}

{{- if .Values.images -}}
    {{- if .Values.images.pullSecrets -}}
        {{- $list = append $list .Values.images.pullSecrets -}}
    {{- end -}}
{{- end -}}

{{-  include "newrelic.common.images.renderPullSecrets" ( dict "pullSecrets" $list "context" .) }}

{{- end -}}
```

This for example passes the tests

Reply:
You are completely right.

Reply:
Fixed in [#fb54ba2](https://github.com/newrelic/nri-kube-events/pull/117/commits/fb54ba2bbfbcf53299deb5ec7a6670c031d95f86)